### PR TITLE
[nnfw/single] enable nnfw for single API

### PR DESCRIPTION
**Pull request by kparichay**
- Update the nnfw framework input model format to work with file path
- Enable nnfw for single API
    - nnfw and tensorflow-lite both support tflite extensions. Added meson option to set default behavior which prioritizes tensorflow-lite
- Add invoke test for nnfw with single API

**Self evaluation:**
1. Build test: [x]Passed [ ]Failed [ ]Skipped
2. Run test: [x]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor pk.kapoor@samsung.com

### CODE REVIEW

File: api/capi/src/nnstreamer-capi-util.c [hunk 1]
```diff
@@ -868,8 +868,34 @@ ml_validate_model_file (const char *model, ml_nnfw_type_e * nnfw)
...
       if (g_str_has_suffix (path_down, ".tflite")) {
-        ml_logi ("The given model [%s] is supposed a tensorflow-lite model.", model);
-        *nnfw = ML_NNFW_TYPE_TENSORFLOW_LITE;
+        /**
+         * .tflite is supported by both tensorflow and nnfw.
+         * Priority decided with ini file.
+         */
+        bool available = FALSE;
+
+#ifdef NNFW_RUNTIME_PRIORITIZE
```
Comment (myungjoo) [inline api/capi/src/nnstreamer-capi-util.c]: I don't recommend to use preprocessors for such decisions. 
It's better with nnstreamer configurations (.ini + envvar + hardcoded-defaults)

Comment (kparichay) [reply api/capi/src/nnstreamer-capi-util.c]: > It's better with nnstreamer configurations (envvar + .ini + hardcoded-defaults)

Updated to this format.

File: api/capi/src/nnstreamer-capi-util.c [hunk 2]
```diff
@@ -868,8 +868,27 @@ ml_validate_model_file (const char *model, ml_nnfw_type_e * nnfw)
...
+
+        ml_check_nnfw_availability (ML_NNFW_TYPE_NNFW, ML_NNFW_HW_ANY,
+            &available_nnfw);
+        ml_check_nnfw_availability (ML_NNFW_TYPE_TENSORFLOW_LITE,
+            ML_NNFW_HW_ANY, &available_tflite);
+
+        if ((nnfw_runtime_priority && available_nnfw) ||
+            (!nnfw_runtime_priority && !available_tflite)) {
+          ml_logi ("The given model [%s] is supposed a nnfw model.", model);
+          *nnfw = ML_NNFW_TYPE_NNFW;
```
Comment (myungjoo) [inline api/capi/src/nnstreamer-capi-util.c]: What if ```!available_tflite && !available_nnfw)```?

Comment (kparichay) [reply api/capi/src/nnstreamer-capi-util.c]: In this function case ML_NNFW_TYPE_ANY, the purpose is to assign a **possible** valid NN framework. Checking for availability is an added bonus I think. 
My assumption is `ml_validate_model_file` is followed by a call to `ml_check_nnfw_availability`.
The case of `!available_tflite && !available_nnfw` will be covered by that.

So, to make this better, I have added a check of `ml_check_nnfw_availibility` at the end of `ML_NNFW_TYPE_ANY` in `ml_validate_model_file` to check if the assigned nnfw is even available or not for all the ext. This covers the case of `!available_tflite && !available_nnfw`, and for other cases of `ML_NNFW_TYPE_ANY`.

File: tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp [hunk 3]
```diff
@@ -41,35 +41,45 @@ TEST (nnstreamer_nnfw_runtime_raw_functions, open_close_00_n)
...
 
-  test_model = g_build_filename (model_path, "add.tflite", NULL);
-  ASSERT_TRUE (g_file_test (test_model, G_FILE_TEST_EXISTS));
-  g_free (test_model);
+  meta_file = g_build_filename (model_path, "metadata", "MANIFEST", NULL);
+  g_return_val_if_fail (g_file_test (meta_file, G_FILE_TEST_EXISTS), FALSE);
+  g_free (meta_file);
+
+  model_file = g_build_filename (model_path, "add.tflite", NULL);
+  g_return_val_if_fail (g_file_test (model_file, G_FILE_TEST_EXISTS), FALSE);
```
Comment (dongju-chae) [inline tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp]: Missing `g_free (meta_file);` ?

Comment (jaeyun-jung) [reply tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp]: line 59 and 63 will report svace issue (no-free function: meta_file and model_file)

Comment (dongju-chae) [reply tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp]: Ah, What I meant is line 64: `g_free (model_path)` --> `g_free (model_file)`

**Comment by myungjoo** [comment]
Without this commit, we still have the same nnfw-unit-test failures:
```
[  201s] + ./tests/tizen_nnfw_runtime/unittest_nnfw_runtime_raw --gst-plugin-path=. --gtest_output=xml:unittest_nnfw_runtime_raw.xml
[  201s] [==========] Running 5 tests from 1 test case.
[  201s] [----------] Global test environment set-up.
[  201s] [----------] 5 tests from nnstreamer_nnfw_runtime_raw_functions
[  201s] [ RUN      ] nnstreamer_nnfw_runtime_raw_functions.check_existence
[  201s] [       OK ] nnstreamer_nnfw_runtime_raw_functions.check_existence (26 ms)
[  201s] [ RUN      ] nnstreamer_nnfw_runtime_raw_functions.open_close_00_n
[  201s] invalid nnpackge directory: null.nnfw
[  201s] Cannot load the model file: null.nnfw[       OK ] nnstreamer_nnfw_runtime_raw_functions.open_close_00_n (2 ms)
[  201s] [ RUN      ] nnstreamer_nnfw_runtime_raw_functions.open_close_01_n
[  201s] nnfw_close called without proper nnfw_openCan't load libOpenCL.so: libOpenCL.so: cannot open shared object file: No such file or directory
[  201s] Can't load libGLES_mali.so: libGLES_mali.so: cannot open shared object file: No such file or directory
[  201s] Can't load libmali.so: libmali.so: cannot open shared object file: No such file or directory
[  201s] Couldn't find any OpenCL library.
[  201s] Error during model prepare
[  201s] nnfw-runtime cannot prepare the session for /home/abuild/rpmbuild/BUILD/nnstreamer-1.0.0/tests/test_models/models../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:82: Failure
[  201s]       Expected: ret
[  201s]       Which is: -22
[  201s] To be equal to: 0
[  201s] nnfw_close called without proper nnfw_opennnfw_close called without proper nnfw_open[  FAILED  ] nnstreamer_nnfw_runtime_raw_functions.open_close_01_n (77 ms)
[  201s] [ RUN      ] nnstreamer_nnfw_runtime_raw_functions.get_dimension
[  201s] 
[  201s] ** (process:31036): CRITICAL **: nnfw_tensors_info_get: assertion 'pdata != NULL' failed
[  201s] 
[  201s] ** (process:31036): CRITICAL **: nnfw_tensors_info_get: assertion 'pdata != NULL' failed
[  201s] Error during model prepare
[  201s] nnfw-runtime cannot prepare the session for /home/abuild/rpmbuild/BUILD/nnstreamer-1.0.0/tests/test_models/models../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:130: Failure
[  201s]       Expected: ret
[  201s]       Which is: -22
[  201s] To be equal to: 0
[  201s] 
[  201s] ** (process:31036): CRITICAL **: nnfw_tensors_info_get: assertion 'info != NULL' failed
[  201s] 
[  201s] ** (process:31036): CRITICAL **: nnfw_tensors_info_get: assertion 'pdata != NULL' failed
[  201s] ../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:143: Failure
[  201s]       Expected: ret
[  201s]       Which is: -22
[  201s] To be equal to: 0
[  201s] ../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:146: Failure
[  201s]       Expected: res.info[0].type
[  201s]       Which is: 1
[  201s] To be equal to: info.info[0].type
[  201s]       Which is: 7
[  201s] ../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:147: Failure
[  201s]       Expected: res.info[0].dimension[0]
[  201s]       Which is: 0
[  201s] To be equal to: info.info[0].dimension[0]
[  201s]       Which is: 1
[  201s] ../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:148: Failure
[  201s]       Expected: res.info[0].dimension[1]
[  201s]       Which is: 19897488
[  201s] To be equal to: info.info[0].dimension[1]
[  201s]       Which is: 1
[  201s] ../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:149: Failure
[  201s]       Expected: res.info[0].dimension[2]
[  201s]       Which is: 64
[  201s] To be equal to: info.info[0].dimension[2]
[  201s]       Which is: 1
[  201s] ../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:150: Failure
[  201s]       Expected: res.info[0].dimension[3]
[  201s]       Which is: 10488744
[  201s] To be equal to: info.info[0].dimension[3]
[  201s]       Which is: 1
[  201s] 
[  201s] ** (process:31036): CRITICAL **: nnfw_tensors_info_get: assertion 'info != NULL' failed
[  201s] 
[  201s] ** (process:31036): CRITICAL **: nnfw_tensors_info_get: assertion 'pdata != NULL' failed
[  201s] ../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:155: Failure
[  201s]       Expected: ret
[  201s]       Which is: -22
[  201s] To be equal to: 0
[  201s] ../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:158: Failure
[  201s]       Expected: res.info[0].type
[  201s]       Which is: 1
[  201s] To be equal to: info.info[0].type
[  201s]       Which is: 7
[  201s] ../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:159: Failure
[  201s]       Expected: res.info[0].dimension[0]
[  201s]       Which is: 0
[  201s] To be equal to: info.info[0].dimension[0]
[  201s]       Which is: 1
[  201s] ../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:160: Failure
[  201s]       Expected: res.info[0].dimension[1]
[  201s]       Which is: 19897488
[  201s] To be equal to: info.info[0].dimension[1]
[  201s]       Which is: 1
[  201s] ../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:161: Failure
[  201s]       Expected: res.info[0].dimension[2]
[  201s]       Which is: 64
[  201s] To be equal to: info.info[0].dimension[2]
[  201s]       Which is: 1
[  201s] ../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:162: Failure
[  201s]       Expected: res.info[0].dimension[3]
[  201s]       Which is: 10488744
[  201s] To be equal to: info.info[0].dimension[3]
[  201s]       Which is: 1
[  201s] nnfw_close called without proper nnfw_open[  FAILED  ] nnstreamer_nnfw_runtime_raw_functions.get_dimension (7 ms)
[  201s] [ RUN      ] nnstreamer_nnfw_runtime_raw_functions.invoke
[  201s] 
[  201s] ** (process:31036): CRITICAL **: nnfw_tensor_memory_set: assertion 'pdata != NULL' failed
[  201s] Error during model prepare
[  201s] nnfw-runtime cannot prepare the session for /home/abuild/rpmbuild/BUILD/nnstreamer-1.0.0/tests/test_models/models../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:213: Failure
[  201s]       Expected: ret
[  201s]       Which is: -22
[  201s] To be equal to: 0
[  201s] 
[  201s] ** (process:31036): CRITICAL **: nnfw_tensor_memory_set: assertion 'mem != NULL' failed
[  201s] 
[  201s] ** (process:31036): CRITICAL **: nnfw_tensor_memory_set: assertion 'pdata != NULL' failed
[  201s] 
[  201s] ** (process:31036): CRITICAL **: nnfw_tensor_memory_set: assertion 'mem != NULL' failed
[  201s] 
[  201s] ** (process:31036): CRITICAL **: nnfw_tensor_memory_set: assertion 'pdata != NULL' failed
[  201s] ../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:225: Failure
[  201s]       Expected: ret
[  201s]       Which is: -22
[  201s] To be equal to: 0
[  201s] ../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:226: Failure
[  201s]       Expected: *((float *)output.data)
[  201s]       Which is: 1.93637e-38
[  201s] To be equal to: 12.0
[  201s]       Which is: 12
[  201s] 
[  201s] ** (process:31036): CRITICAL **: nnfw_tensor_memory_set: assertion 'pdata != NULL' failed
[  201s] ../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:230: Failure
[  201s]       Expected: ret
[  201s]       Which is: -22
[  201s] To be equal to: 0
[  201s] ../tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp:231: Failure
[  201s]       Expected: *((float *)output.data)
[  201s]       Which is: 1.93637e-38
[  201s] To be equal to: 3.0
[  201s]       Which is: 3
[  201s] nnfw_close called without proper nnfw_open[  FAILED  ] nnstreamer_nnfw_runtime_raw_functions.invoke (6 ms)
[  201s] [----------] 5 tests from nnstreamer_nnfw_runtime_raw_functions (119 ms total)
[  201s] 
[  201s] [----------] Global test environment tear-down
[  201s] [==========] 5 tests from 1 test case ran. (121 ms total)
[  201s] [  PASSED  ] 2 tests.
[  201s] [  FAILED  ] 3 tests, listed below:
[  201s] [  FAILED  ] nnstreamer_nnfw_runtime_raw_functions.open_close_01_n
[  201s] [  FAILED  ] nnstreamer_nnfw_runtime_raw_functions.get_dimension
[  201s] [  FAILED  ] nnstreamer_nnfw_runtime_raw_functions.invoke
[  201s] 
[  201s]  3 FAILED TESTS
[  201s] error: Bad exit status from /var/tmp/rpm-tmp.scBy7a (%build)
```

Thus, I won't stop this PR with this unittest fails. However, we do have correctness issues with nnfw-runtime subplugin.

File: tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp (hunk 3 above)
Comment (kparichay) [reply tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp]: updated.

**Comment by kparichay** [comment]
@myungjoo As you said, this PR will not be effected by unittest fail. However, I wanted to point out that the unittest fail is not because of our implementation but nnfw package (https://github.com/nnsuite/nnstreamer/issues/1903#issuecomment-564379433). 
I have tested the nnfw extension for nnstreamer with armv7l and aarch64 on my local machine (with hotfix to nnfw) and all unittests pass.

**Pull request merged**